Meraki util - Add method to encode parameters in the URL

kbreit opened the pull request:
##### SUMMARY
Some of the Meraki endpoints require parameters to be passed via the URL instead of in the body. A new method was created in the module utility format the parameter string and append to the URL. Also, a new parameter was added to the construct_path() method to allow passing the parameters, which are handled automatically.

Note, I am not actually doing any real encoding, just concatenating. If there is an Ansible method which already does this, please let me know and I'll use that instead.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
meraki

##### VERSION
ansible 2.8.0.dev0 (meraki/util_url_params cb6cdf4dab) last updated 2018/11/21 21:28:56 (GMT -500)
  config file = None
  configured module search path = ['/Users/kbreit/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/kbreit/Documents/Programming/ansible/lib/ansible
  executable location = /Users/kbreit/Documents/Programming/ansible/bin/ansible
  python version = 3.5.4 (default, Feb 25 2018, 14:56:02) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]

kbreit committed:
Add new method to encode parameters in the URL
- I'm not really encoding, I'm sure this is broke
- There maybe an Ansible native way to do this

kbreit committed:
Fix whitepace

dagwieers commented:
I wrote my own "query-string" handling functions (see ACI and MSC module_utils). I do think you need to encode the keys and values to avoid any syntax issues.

dagwieers commented:
Here they are:

https://github.com/ansible/ansible/blob/9fb4e24660ff72754fcd190c960b904f7e53da34/lib/ansible/module_utils/network/aci/aci.py#L388-L396

https://github.com/ansible/ansible/blob/9fb4e24660ff72754fcd190c960b904f7e53da34/lib/ansible/module_utils/network/aci/msc.py#L80-L83

But looking at this now, you probably want to use urlencode() which is currently not available from urls.py modutils. I guess we need to add that.

kbreit committed:
Added urlencode support
- Relies on urllib module
- Fixed string delimiter

kbreit committed:
Enable URL params
- construct_params_list() creates a list of parameters to encode
- encode_url_params() does encoding in a simple manner

kbreit commented:
> But looking at this now, you probably want to use urlencode() which is currently not available from urls.py modutils. I guess we need to add that.

Fixed.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -249,7 +261,24 @@ def get_template_id(self, name, data):
+    def encode_url_params(self, params):
+        """Encodes key value pairs for URL"""
+        param_string = "?"
+        for i, (k, v) in enumerate(params.items()):
```
This is what `urlencode()` is already doing for you., so the function would be:
```python
    def encode_url_params(self, params):
        """Encodes key value pairs for URL"""
        return '?' + urlencode(params.items())

kbreit replied:
Well that's not good. I had fixed this, committed, but it seems like the data got lost somewhere. Need to rewrite :/

kbreit replied:
Okay, fixed.

kbreit committed:
Added proper methods for urlencoding

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -152,6 +153,17 @@ def is_update_required(self, original, proposed, optional_ignore=None):
                     is_changed = True
         return is_changed
 
+    def construct_params_list(self, keys, aliases=None):
```
I see this function twice in this commit?

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -152,6 +153,17 @@ def is_update_required(self, original, proposed, optional_ignore=None):
+            args_list.append(item)
+        return args_list
+
+    def encode_url_params(self, params):
```
I see this function twice in this commit?

kbreit replied:
I had it in there and rewrote it thinking it wasn't there. Fixed.

kbreit committed:
Remove duplicate functions

kbreit committed:
Remove blank line for PEP8

kbreit commented:
@dagwieers Dag, Shippable is happy with the current version and I think I addressed your comments. Please review when you can.
